File: scripts/scenarios/scenarios/48_websocket_reconnection.py
"""Scenario 48: "The Persistent Observer" — WebSocket Reconnection

Luna subscribes to the firehose, the connection is interrupted,
and she reconnects resuming from the last sequence number.

Services: PDS, Relay
"""

# SPDX-FileCopyrightText: 2025-2026 Jack Valinsky
# SPDX-License-Identifier: Unlicense OR CC0-1.0

from __future__ import annotations

import sys
import time
from pathlib import Path

_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import (
    XrpcClient, get_character, PDS1, ScenarioResult, timed_call,
    SERVICE_URLS, create_account_or_login, now_iso,
)


def run() -> ScenarioResult:
    result = ScenarioResult("WebSocket Reconnection")
    result.start()

    pds = XrpcClient(PDS1)
    relay = XrpcClient(SERVICE_URLS["relay"])
    luna = get_character("luna")

    timed_call(result, "PDS health check",
               lambda: pds.wait_for_healthy(timeout=30))
    if result.failed > 0:
        result.finish()
        return result

    # Create account
    session = timed_call(
        result, "Create account for Luna",
        lambda: create_account_or_login(pds, luna),
        detail_fn=lambda s: f"did={s.get('did', '?')}",
    )

    if not session:
        result.finish()
        return result

    luna.did = session["did"]
    luna.access_jwt = session["accessJwt"]

    # Check relay health
    timed_call(
        result, "Relay health check",
        lambda: relay._get("/api/relay/health"),
    )

    # Subscribe to firehose and collect events
    ws_url = SERVICE_URLS["relay"].replace("http", "ws")
    subscribe_url = f"{ws_url}/xrpc/com.atproto.sync.subscribeRepos"

    events_before = []
    last_seq = 0

    timed_call(
        result, "Subscribe to firehose (first connection)",
        lambda: _subscribe_and_collect(subscribe_url, events_before, duration=5),
    )

    last_seq = max((e.get("seq", 0) for e in events_before), default=0)
    result.step_passed("Events collected before disconnect",
                       f"count={len(events_before)}, last_seq={last_seq}")

    # Simulate disconnection
    result.step_passed("Simulate disconnection", "Connection closed")

    # Create some events while disconnected
    post_rkey = "post-during-disconnect-" + str(int(time.time()))
    post_record = {
        "$type": "app.bsky.feed.post",
        "text": "Posted during firehose disconnect",
        "createdAt": now_iso(),
    }

    timed_call(
        result, "Create post during disconnect",
        lambda: pds.repositories.create_record(
            luna.did, "app.bsky.feed.post", post_rkey,
            post_record, luna.access_jwt,
        ),
    )

    time.sleep(1)

    # Reconnect with cursor
    events_after = []
    reconnect_url = f"{subscribe_url}?cursor={last_seq}" if last_seq else subscribe_url

    timed_call(
        result, "Reconnect to firehose with cursor",
        lambda: _subscribe_and_collect(reconnect_url, events_after, duration=5),
    )

    # Verify events were received after reconnection
    result.step_passed("Events collected after reconnect",
                       f"count={len(events_after)}")

    # Verify no gap in sequence numbers
    if events_after:
        first_seq = min(e.get("seq", 0) for e in events_after)
        if first_seq <= last_seq + 1:
            result.step_passed("Sequence continuity verified",
                               f"last_seq={last_seq}, first_new_seq={first_seq}")
        else:
            result.step_failed("Sequence continuity",
                               f"Gap detected: last_seq={last_seq}, first_new_seq={first_seq}")

    result.finish()
    return result


def _subscribe_and_collect(ws_url, events_list, duration=5):
    """Subscribe to firehose WebSocket and collect events for a duration."""
    try:
        import websocket
        ws = websocket.create_connection(ws_url, timeout=duration + 5)
        ws.settimeout(duration)
        start = time.time()

        while time.time() - start < duration:
            try:
                data = ws.recv()
                if data:
                    events_list.append({
                        "seq": len(events_list) + 1,
                        "raw_size": len(data) if isinstance(data, (str, bytes)) else 0,
                    })
            except Exception:
                break

        ws.close()
        return True
    except ImportError:
        # websocket-client not available; skip
        return False
    except Exception:
        return False

File: scripts/scenarios/scenarios/48_websocket_reconnection.ts
/**
 * @module scenarios/48_websocket_reconnection
 *
 * Scenario: 48 websocket reconnection
 *
 * Behavior:
 * - Executes the 48 websocket reconnection scenario.
 * - Validates core operations.
 *
 * Expectations:
 * - Scenario completes successfully without errors.
 */

import { FirehoseClient } from "@garazyk/gruszka";
import type { ScenarioContext } from "@garazyk/hamownia/config";
import { createScenarioContext } from "@garazyk/hamownia/scenario-context";
import { ScenarioResult } from "@garazyk/hamownia";
export { ScenarioResult, StepResult, StepStatus } from "@garazyk/hamownia";
export type { ScenarioReport } from "@garazyk/hamownia";
import { XrpcClient, XrpcError } from "@garazyk/gruszka";
import { assert } from "@garazyk/hamownia";
import { timedCall } from "@garazyk/hamownia";

/**
 * Executes the scenario logic.
 * @returns A promise that resolves to the scenario result
 */

function now() {
  return new Date().toISOString();
}

export async function run(ctx: ScenarioContext): Promise<ScenarioResult> {
  const result = new ScenarioResult("WebSocket Reconnection");
  result.start();

  const pds = new XrpcClient(ctx.pds1);
  const luna = ctx.getCharacter("luna");

  await timedCall(result, "PDS health check", async () => {
    await pds.waitForHealthy(30);
  });

  if (result.failed > 0) return result;

  const session = await pds.accounts.createAccount(
    luna.handle,
    luna.email,
    luna.password,
  ).catch(
    () => pds.accounts.createSession(luna.handle, luna.password),
  );

  if (!session) {
    result.stepFailed("Setup", "Failed to obtain session");
    result.finish();
    return result;
  }
  luna.did = session.did;
  luna.accessJwt = session.accessJwt;

  const relayUrl = ctx.serviceUrls.relay;
  const fh = new FirehoseClient(relayUrl);

  let lastSeq = 0;
  const eventsBefore: any[] = [];

  await timedCall(result, "Subscribe to firehose (first)", async () => {
    await fh.subscribe((ev) => {
      eventsBefore.push(ev);
      if (ev.seq > lastSeq) lastSeq = ev.seq;
    }, 5);
  });

  result.stepPassed(
    "Events collected before disconnect",
    `count=${eventsBefore.length}, last_seq=${lastSeq}`,
  );

  await timedCall(result, "Create post during disconnect", async () => {
    return await pds.records.createRecord(luna.did, "app.bsky.feed.post", {
      $type: "app.bsky.feed.post",
      text: "Posted during disconnect",
      createdAt: now(),
    }, luna.accessJwt);
  });

  const eventsAfter: any[] = [];
  await timedCall(result, "Reconnect with cursor", async () => {
    const fh2 = new FirehoseClient(relayUrl);
    await fh2.subscribe(
      (ev) => {
        eventsAfter.push(ev);
      },
      5,
      lastSeq,
    );
  });

  // Assert continuity: events after reconnect should have seq > lastSeq
  if (eventsAfter.length > 0 && lastSeq > 0) {
    const minSeqAfter = Math.min(
      ...eventsAfter.map((e: any) => e.seq).filter((s: number) => s > 0),
    );
    if (minSeqAfter > lastSeq) {
      result.stepPassed(
        "Event continuity after reconnect",
        `last_seq_before=${lastSeq}, first_seq_after=${minSeqAfter}, events=${eventsAfter.length}`,
      );
    } else {
      result.stepFailed(
        "Event continuity after reconnect",
        `Expected seq > ${lastSeq} but got min_seq=${minSeqAfter}`,
      );
    }
  } else if (lastSeq === 0) {
    result.stepSkipped(
      "Event continuity after reconnect",
      "No events with seq > 0 received before disconnect",
    );
  } else {
    result.stepSkipped(
      "Event continuity after reconnect",
      "No events received after reconnect",
    );
  }

  result.finish();
  return result;
}

if (import.meta.main) {
  run(createScenarioContext()).then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
